### PR TITLE
Add .n to lazy, fixes #39

### LINKED ISSUE #39: Should there be an option to make lazy follow symbols n times?
Because the `data` argument in  `lazy_eval()` allows you to look in a data.frame/list before the environment, it is not always the case that you'd like lazy to follow promises all the way until they can be evaluated in the global environment. 

For example, if I wanted to change the `select()` function to accept named arguments, `lazy()` will follow the promises too far. Setting `.follow_symbols = F` doesn't work either, because it checks the names from the function in data, not what the user input.
```r
library(lazyeval)
library(dplyr)

# dplyr::select
# select <- function(.data, ...) {
#   select_(.data, .dots = lazyeval::lazy_dots(...))
# }

# direct translation of lazy_dots() to lazy()
select2 <- function(.data, group1, group2) {
  out1 <- select_(.data, .dots = lazy(group1))
  out2 <- select_(.data, .dots = lazy(group2))
  
  list(out1, out2)
}

select3 <- function(.data, group1, group2) {
  out1 <- select_(.data, .dots = lazy(group1, .follow_symbols = F))
  out2 <- select_(.data, .dots = lazy(group2, .follow_symbols = F))
  
  list(out1, out2)
}

# select2 fails when a variable name is technically a promise 
# (for example if it's a function in another package)
data.frame(x = 1, order = 2) %>% select2(x, order)
#>  Error in lazy(group2) : Promise has already been forced 

# select3 doesn't do what I want either
data.frame(x = 1, order = 2) %>% select3(x, order)
#>  Error in eval(expr, envir, enclos) : object 'x' not found 
```

But if we could control the number of times that promises are followed, then we could do this:
```R
select4 <- function(.data, group1, group2) {
  out1 <- select_(.data, .dots = lazy(group1, .n = 1))
  out2 <- select_(.data, .dots = lazy(group2, .n = 1))
  
  list(out1, out2)
}
data.frame(x = 1, order = 2) %>% select4(x, order)
```

I know that this is related to #30, but that pull request didn't fix my examples. I'll submit a rough PR that has .n, and maybe the two can be combined.

### PR DESCRIPTION
**Pull request by gergness**
First use of Rcpp (and pull request), sorry if there's any slop!

### CODE REVIEW

File: src/lazy.c [hunk 1]
```diff
@@ -1,10 +1,12 @@
...
 #include <Rdefines.h>
 
-SEXP promise_as_lazy(SEXP promise, SEXP env, int follow_symbols) {
+SEXP promise_as_lazy(SEXP promise, SEXP env, int follow_symbols, int n) {
   // recurse until we find the real promise, not a promise of a promise
   // never go past the global environment
-  while(TYPEOF(promise) == PROMSXP && env != R_GlobalEnv) {
+  // don't go more than n times
+  int i = 0;
+  while(TYPEOF(promise) == PROMSXP && env != R_GlobalEnv && (n == NA_INTEGER || i <= n)) {
```
Comment (robertzk) [inline src/lazy.c]: This will cap it at  `n+1` times instead of `n` times, which is incongruous with the comment. Was that the intention or is it irrelevant?

Comment (robertzk) [inline src/lazy.c]: An easy fix is to let `int i = 1;` in the previous line.

File: src/lazy.c [hunk 2]
```diff
@@ -15,12 +17,14 @@ SEXP promise_as_lazy(SEXP promise, SEXP env, int follow_symbols) {
     // If the promise is threaded through multiple functions, we'll
     // get some symbols along the way. If the symbol is bound to a promise
     // keep going on up
-    if (follow_symbols && TYPEOF(promise) == SYMSXP) {
+    // Unless we have reached n loops, then we should return the symbol
+    if (follow_symbols && TYPEOF(promise) == SYMSXP && (n == NA_INTEGER || i < n)) {
```
Comment (robertzk) [inline src/lazy.c]: Oh... nevermind! Kind of weird but I guess that works.

File: R/lazy.R [hunk 3]
```diff
@@ -14,6 +14,8 @@
 #' @param env Environment in which to evaluate expr.
 #' @param .follow_symbols If \code{TRUE}, the default, follows promises across
 #'   function calls. See \code{vignette("chained-promises")} for details.
+#' @param .n If \code{.follow_symbols = TRUE} will follow promises .n times.
```
Comment (robertzk) [inline R/lazy.R]: Can you change `.n times` to `\code{.n} times`, please?

File: R/lazy.R [hunk 4]
```diff
@@ -14,6 +14,8 @@
...
+#'   NA to follow until it finds a non-promise.
```
Comment (robertzk) [inline R/lazy.R]: This line is not 100% clear to me upon an immediate read. Can you rephrase slightly?

Comment (robertzk) [inline R/lazy.R]: Why not use `Inf` as the default instead of `NA` as the "infinite chain" bound? This would make more sense to me.

Comment (hadley) [inline R/lazy.R]: The argument seems redundant with `.follow_symbols` to me - why not make `.follow_symbols` take an integer, or true or false.

Comment (robertzk) [inline R/lazy.R]: That's another great point. ^ That sounds like even better design.

Comment (gergness) [inline R/lazy.R]: Was trying to avoid breaking existing code, but that makes sense. I've updated. Thanks!

File: src/lazy.c (hunk 1 above)
Comment (gergness) [inline src/lazy.c]: Thanks! I'm not sure that line did what I thought it did -- it ended up causing a bug in lazy_dots() when I moved .n into .follow_symbols

File: R/lazy.R (hunk 4 above)
Comment (gergness) [inline R/lazy.R]: Internally, I'm still using NA_INTEGER as the TRUE value, because this was the best way I could think to do this (I'm not used to static types), let me know if you'd like to do it another way.

Comment (gergness) [inline R/lazy.R]: And finally @robertzk, I added a bit to the vignette to try to clarify the argument -- do you think it's clear enough of an explanation? Thanks again!

File: R/lazy-dots.R [hunk 5]
```diff
@@ -32,7 +32,7 @@ lazy_dots <- function(..., .follow_symbols = FALSE) {
   if (nargs() == 0 || (nargs() == 1 &&  ! missing(.follow_symbols))) {
     return(structure(list(), class = "lazy_dots"))
   }
-
+  if (isTRUE(.follow_symbols)) .follow_symbols <- NA
```
Comment (hadley) [inline R/lazy-dots.R]: I prefer to always wrap the body of an if with `{}` and put on its own line.

Comment (gergness) [inline R/lazy-dots.R]: Yep - no problem

**Comment by gergness** [comment]
Sorry - I didn't have #30 pulled correctly to my computer -- it actually does fix #39
